armory/scenarios/poison.py

```diff
@@ -518,28 +532,116 @@ def evaluate_current(self):
-                self.poison_index,
-                self.indices_to_keep,
-                # Get unpoisoned test set
-                config_loading.load_dataset(
-                    self.config["dataset"],
-                    split="test",
-                    num_batches=None,
-                    **self.dataset_kwargs,
-                ),
+    def init_explanatory(self):
+        self.test_set_class_labels = set()
+        self.test_x = []
+        self.test_y = []
+        self.test_y_pred_class = []
+
+    def run_explanatory(self):
+        self.hub.set_context(stage="explanatory")
+        self.test_set_class_labels.update(self.y)
+        self.test_x.append(self.x)
+        self.test_y.extend(self.y)
+        self.test_y_pred_class.extend(self.y_pred.argmax(axis=1))
+
+    def finalize_explanatory(self):
+        self.test_x = np.concatenate(self.test_x)
+        self.test_y = np.array(self.test_y)
+        self.test_y_pred_labels = np.array(self.test_y_pred_class)
+        self.test_set_class_labels = sorted(self.test_set_class_labels)
+
+    def get_train_majority_mask_and_ceilings(self):
+        """
+        get majority ceilings on unpoisoned part of train set
+        """
+        if self.explanatory_model is None:
+            raise ValueError("No explanatory model")
+
+        class_majority_mask = metrics.get("class_majority_mask")
+        activations = self.explanatory_model.get_activations(
+            self.x_poison[~self.poisoned]
+        )
+        (
+            self.majority_mask_train_unpoisoned,
+            self.majority_ceilings,
+        ) = class_majority_mask(
+            activations,
+            self.y_poison[~self.poisoned],
+        )
+        return self.majority_mask_train_unpoisoned, self.majority_ceilings
+
+    def get_test_majority_mask(self):
+        """
+        get majority ceilings on unpoisoned part of test set
+        """
+        if self.explanatory_model is None:
+            raise ValueError("No explanatory model")
+        if not hasattr(self, "majority_ceilings"):
+            raise ValueError("Must first call 'get_train_majority_mask_and_ceilings'")
+        class_majority_mask = metrics.get("class_majority_mask")
+        activations = self.explanatory_model.get_activations(self.test_x)
+        # use copy of majority ceilings computed from train set
+        self.majority_mask_test_set, _ = class_majority_mask(
+            activations,
+            self.test_y,
+            majority_ceilings=copy.copy(self.majority_ceilings),
+        )
+
+        return self.majority_mask_test_set
+
+    def record_bias_results(self, chi2_spd, class_ids, record_prefix, log_prefix):
+        for class_id in class_ids:
+            chi2, spd = chi2_spd[class_id]
+            self.hub.record(
+                f"{record_prefix}bias_chi^2_p_value_{str(class_id).zfill(2)}", chi2
+            )
+            self.hub.record(f"{record_prefix}bias_spd_{str(class_id).zfill(2)}", spd)
+        log.info(
```

Comment:
looks like these log lines should be indented under the for loop

Reply:
Removed.